Dyn trait causes `values of this type cannot be used as base sorted instances`

The following code:

```rust
use flux_rs::*;

#[generics(Self as base)]
trait Trait {}

fn foo(x: &dyn Trait) {}
```

gives:

```
error[E0999]: values of this type cannot be used as base sorted instances
 --> attic/playground.rs:6:16
  |
6 | fn foo(x: &dyn Trait) {}
  |                ^^^^^
-Ztrack-diagnostics: created at crates/flux-fhir-analysis/src/conv/mod.rs:1455:26
```

--- a/crates/flux-desugar/src/desugar.rs
+++ b/crates/flux-desugar/src/desugar.rs
@@ -245,7 +245,6 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
                     return Err(self.emit_err(errors::UnresolvedGenericParam::new(param.name)));
                 };
 
-                // The default type is wrong for extern specs
                 let maybe_extern = self.genv.maybe_extern_id(def_id);
                 let kind = match &param.kind {
                     surface::GenericParamKind::Type => {

--- a/crates/flux-fhir-analysis/src/conv/mod.rs
+++ b/crates/flux-fhir-analysis/src/conv/mod.rs
@@ -1445,14 +1445,17 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
     /// Convert an [`rty::Ty`] into a [`rty::GenericArg::Base`] if possible or raise an error
     /// if the type cannot be converted into a [`rty::SubsetTy`].
     fn ty_to_base_generic(&self, span: Span, ty: &rty::Ty) -> QueryResult<rty::GenericArg> {
-        let ctor = ty
-            .shallow_canonicalize()
-            .to_subset_ty_ctor()
-            .ok_or_else(|| {
-                self.genv
-                    .sess()
-                    .emit_err(errors::InvalidBaseInstance::new(span))
-            })?;
+        let ctor = if ty == &rty::Ty::trait_object_dummy_self() {
+            rty::SubsetTyCtor::trait_object_dummy_self()
+        } else {
+            ty.shallow_canonicalize()
+                .to_subset_ty_ctor()
+                .ok_or_else(|| {
+                    self.genv
+                        .sess()
+                        .emit_err(errors::InvalidBaseInstance::new(span))
+                })?
+        };
         Ok(rty::GenericArg::Base(ctor))
     }
 

--- a/crates/flux-middle/src/queries.rs
+++ b/crates/flux-middle/src/queries.rs
@@ -681,7 +681,7 @@ impl<'genv, 'tcx> Queries<'genv, 'tcx> {
 
 /// [Resolve] the `def_id` and *dispatch* it to a provider (`local`, `external`, or `default`).
 ///
-/// [Resolve]: resolve_id
+/// [Resolve]: GlobalEnv::resolve_id
 fn dispatch_query<R>(
     genv: GlobalEnv,
     def_id: DefId,

--- a/crates/flux-middle/src/rty/canonicalize.rs
+++ b/crates/flux-middle/src/rty/canonicalize.rs
@@ -167,7 +167,7 @@ impl CanonicalTy {
                     let constr = SubsetTy::new(
                         bty.clone(),
                         Expr::nu(),
-                        Expr::and(constr.pred.clone(), Expr::eq(Expr::nu(), idx)),
+                        Expr::and(&constr.pred, Expr::eq(Expr::nu(), idx)),
                     );
                     Some(Binder::bind_with_sort(constr, sort))
                 } else {

--- a/crates/flux-middle/src/rty/fold.rs
+++ b/crates/flux-middle/src/rty/fold.rs
@@ -726,6 +726,7 @@ impl TypeSuperVisitable for BaseTy {
                 resume_ty.visit_with(visitor)?;
                 upvars.visit_with(visitor)
             }
+            BaseTy::Dynamic(exi_preds, _) => exi_preds.visit_with(visitor),
             BaseTy::Int(_)
             | BaseTy::Uint(_)
             | BaseTy::Bool
@@ -734,8 +735,8 @@ impl TypeSuperVisitable for BaseTy {
             | BaseTy::Char
             | BaseTy::Closure(_, _, _)
             | BaseTy::Never
+            | BaseTy::Infer(_)
             | BaseTy::Param(_) => ControlFlow::Continue(()),
-            BaseTy::Dynamic(exi_preds, _) => exi_preds.visit_with(visitor),
         }
     }
 }
@@ -761,14 +762,6 @@ impl TypeSuperFoldable for BaseTy {
             BaseTy::Array(ty, c) => {
                 BaseTy::Array(ty.try_fold_with(folder)?, c.try_fold_with(folder)?)
             }
-            BaseTy::Int(_)
-            | BaseTy::Param(_)
-            | BaseTy::Uint(_)
-            | BaseTy::Bool
-            | BaseTy::Float(_)
-            | BaseTy::Str
-            | BaseTy::Char
-            | BaseTy::Never => self.clone(),
             BaseTy::Closure(did, args, gen_args) => {
                 BaseTy::Closure(*did, args.try_fold_with(folder)?, gen_args.clone())
             }
@@ -782,6 +775,15 @@ impl TypeSuperFoldable for BaseTy {
             BaseTy::Dynamic(preds, region) => {
                 BaseTy::Dynamic(preds.try_fold_with(folder)?, region.try_fold_with(folder)?)
             }
+            BaseTy::Int(_)
+            | BaseTy::Param(_)
+            | BaseTy::Uint(_)
+            | BaseTy::Bool
+            | BaseTy::Float(_)
+            | BaseTy::Str
+            | BaseTy::Char
+            | BaseTy::Infer(_)
+            | BaseTy::Never => self.clone(),
         };
         Ok(bty)
     }

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -1293,6 +1293,7 @@ pub enum BaseTy {
     Coroutine(DefId, /*resume_ty: */ Ty, /* upvar_tys: */ List<Ty>),
     Dynamic(List<Binder<ExistentialPredicate>>, Region),
     Param(ParamTy),
+    Infer(TyVid),
 }
 
 impl BaseTy {
@@ -1459,6 +1460,7 @@ impl BaseTy {
             | BaseTy::Coroutine(..)
             | BaseTy::Dynamic(_, _)
             | BaseTy::Never => Sort::unit(),
+            BaseTy::Infer(_) => tracked_span_bug!(),
         }
     }
 
@@ -1526,6 +1528,7 @@ impl<'tcx> ToRustc<'tcx> for BaseTy {
                 // let args = tcx.mk_args_from_iter(args);
                 // ty::Ty::new_generator(*tcx, *def_id, args, mov)
             }
+            BaseTy::Infer(ty_vid) => ty::Ty::new_var(tcx, *ty_vid),
         }
     }
 }
@@ -1551,6 +1554,14 @@ pub type RefineArgs = List<Expr>;
 pub type SubsetTyCtor = Binder<SubsetTy>;
 
 impl SubsetTyCtor {
+    /// This is the subset type version of [`Ty::trait_object_dummy_self`]
+    pub fn trait_object_dummy_self() -> Self {
+        SubsetTyCtor::bind_with_sort(
+            SubsetTy::trivial(BaseTy::Infer(TyVid::from_u32(0)), Expr::nu()),
+            Sort::Infer(SortInfer::SortVar(SortVid::from_u32(0))),
+        )
+    }
+
     pub fn as_bty_skipping_binder(&self) -> &BaseTy {
         &self.as_ref().skip_binder().bty
     }
@@ -1572,26 +1583,28 @@ impl SubsetTyCtor {
 ///
 /// These are mainly found under a [`Binder`] with a single variable of the base type's sort. This
 /// can be interpreted as a type constructor or an existial type. For example, under a binder with a
-/// variable `v` of sort `int`, we can interpret `{i32[v] | v > 0}` as a lambda `λv:int. {i32[v] | v > 0}`
-/// that "constructs" types when applied to ints, or as an existential type `∃v:int. {i32[v] | v > 0}`.
+/// variable `v` of sort `int`, we can interpret `{i32[v] | v > 0}` as:
+/// - A lambda `λv:int. {i32[v] | v > 0}` that "constructs" types when applied to ints, or
+/// - An existential type `∃v:int. {i32[v] | v > 0}`.
+///
 /// This second interpretation is the reason we call this a subset type, i.e., the type `∃v. {b[v] | p}`
-/// corresponds to the subset of values of (base) type `b` whose index satisfies `p`. In other words,
-/// these are the types written as `B{v: p}` in the surface syntax and correspond to the types
-/// supported in other refinement type systems like Liquid Haskell (with the difference that we are
-/// explicit about separating refinements from program values via an index).
+/// corresponds to the subset of values of  type `b` whose index satisfies `p`.  These are the types
+/// written as `B{v: p}` in the surface syntax and correspond to the types supported in other
+/// refinement type systems like Liquid Haskell (with the difference that we are explicit
+/// about separating refinements from program values via an index).
 ///
 /// The main purpose for subset types is to be used as generic arguments of [kind base] when
-/// interpreted as type constructors. A subset type has two key properties that makes them suitable
-/// for that.
+/// interpreted as type constructors. They have two key properties that makes them suitable
+/// for this:
 ///
-/// First, because subset types are syntactically restricted, they make it easier to relate types
-/// structurally (e.g., for subtyping). For instance, given two types `S<λv. T1>` and `S<λ. T2>`,
-/// since we know `T1` and `T2` must be subset types, we also know they match structurally
-/// (at least shallowly). The syntactic restriction also rules out more complex types like
-/// `S<λv. (i32[v], i32[0])>` which simplifies some operations on types.
+/// 1. **Syntacitc Restriction**: Subset types are syntactically restricted, making it easier to
+/// relate types structurally (e.g., for subtyping). For instance, given two types `S<λv. T1>` and
+/// `S<λ. T2>`, if `T1` and `T2` are subset types, we know they match structurally (at least
+/// shallowly). The syntactic restriction also rules out complex types like `S<λv. (i32[v], i32[0])>`,
+/// simplifying some operations on types.
 ///
-/// Second, subset types can be eagerly canonicalized via [*strengthening*] during substitution. For
-/// example, suppose we have a function:
+/// 2. **Eager Canonicalization**: Subset types can be eagerly canonicalized via [*strengthening*]
+/// during substitution. For example, suppose we have a function:
 /// ```text
 /// fn foo<T>(x: T[@a], y: { T[@b] | b == a }) { }
 /// ```
@@ -1610,14 +1623,18 @@ impl SubsetTyCtor {
 /// [*strengthening*]: https://arxiv.org/pdf/2010.07763.pdf
 #[derive(PartialEq, Clone, Eq, Hash, TyEncodable, TyDecodable)]
 pub struct SubsetTy {
-    /// **NOTE:** This [`BaseTy`] is mainly going to be under a [`Binder`]. It is not yet clear whether
+    /// The base type `b` in the subset type `{b[e] | p}`.
+    ///
+    /// **NOTE:** This is mostly going to be under a [`Binder`]. It is not yet clear to me whether
     /// this [`BaseTy`] should be able to mention variables in the binder. In general, in a type
     /// `∃v. {b[e] | p}`, it's fine to mention `v` inside `b`, but since [`SubsetTy`] is meant to
-    /// facilitate syntactic manipulation we may restrict this.
+    /// facilitate syntactic manipulation we may want to restrict this.
     pub bty: BaseTy,
-    /// This can be an arbitrary expression which makes manipulation easier, but since this is mostly
-    /// going to be under a binder we expect it to be [`Expr::nu()`].
+    /// The refinement index `e` in the subset type `{b[e] | p}`. This can be an arbitrary expression,
+    /// which makes manipulation easier. However, since this is mostly found under a binder, we expect
+    /// it to be [`Expr::nu()`].
     pub idx: Expr,
+    /// The predicate `p` in the subset type `{b[e] | p}`.
     pub pred: Expr,
 }
 

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -1598,26 +1598,26 @@ impl SubsetTyCtor {
 /// for this:
 ///
 /// 1. **Syntacitc Restriction**: Subset types are syntactically restricted, making it easier to
-/// relate types structurally (e.g., for subtyping). For instance, given two types `S<λv. T1>` and
-/// `S<λ. T2>`, if `T1` and `T2` are subset types, we know they match structurally (at least
-/// shallowly). The syntactic restriction also rules out complex types like `S<λv. (i32[v], i32[0])>`,
-/// simplifying some operations on types.
+///    relate types structurally (e.g., for subtyping). For instance, given two types `S<λv. T1>` and
+///    `S<λ. T2>`, if `T1` and `T2` are subset types, we know they match structurally (at least
+///    shallowly). The syntactic restriction also rules out complex types like `S<λv. (i32[v], i32[0])>`,
+///    simplifying some operations on types.
 ///
 /// 2. **Eager Canonicalization**: Subset types can be eagerly canonicalized via [*strengthening*]
-/// during substitution. For example, suppose we have a function:
-/// ```text
-/// fn foo<T>(x: T[@a], y: { T[@b] | b == a }) { }
-/// ```
-/// If we instantiate `T` with `λv. { i32[v] | v > 0}`, after substitution and applying the lambda
-/// (the indexing syntax `T[a]` corresponds to an application of the lambda), we get:
-/// ```text
-/// fn foo(x: {i32[@a] | a > 0}, y: { { i32[@b] | b > 0 } | b == a }) { }
-/// ```
-/// Via *strengthening* we can canonicalize this to
-/// ```text
-/// fn foo(x: {i32[@a] | a > 0}, y: { i32[@b] | b == a && b > 0 }) { }
-/// ```
-/// As a result, we can guarantee the syntactic restriction through substitution.
+///    during substitution. For example, suppose we have a function:
+///    ```text
+///    fn foo<T>(x: T[@a], y: { T[@b] | b == a }) { }
+///    ```
+///    If we instantiate `T` with `λv. { i32[v] | v > 0}`, after substitution and applying the lambda
+///    (the indexing syntax `T[a]` corresponds to an application of the lambda), we get:
+///    ```text
+///    fn foo(x: {i32[@a] | a > 0}, y: { { i32[@b] | b > 0 } | b == a }) { }
+///    ```
+///    Via *strengthening* we can canonicalize this to
+///    ```text
+///    fn foo(x: {i32[@a] | a > 0}, y: { i32[@b] | b == a && b > 0 }) { }
+///    ```
+///    As a result, we can guarantee the syntactic restriction through substitution.
 ///
 /// [kind base]: GenericParamDefKind::Base
 /// [*strengthening*]: https://arxiv.org/pdf/2010.07763.pdf

--- a/crates/flux-middle/src/rty/pretty.rs
+++ b/crates/flux-middle/src/rty/pretty.rs
@@ -316,8 +316,8 @@ impl Pretty for Ty {
                     join!(", ", &alias_ty.refine_args)
                 )
             }
-            TyKind::Infer(_) => {
-                w!("□")
+            TyKind::Infer(ty_vid) => {
+                w!("{ty_vid:?}")
             }
         }
     }
@@ -458,6 +458,9 @@ impl Pretty for BaseTy {
             }
             BaseTy::Dynamic(preds, re) => {
                 w!("dyn {:?} + {:?}", join!(" + ", preds), re)
+            }
+            BaseTy::Infer(ty_vid) => {
+                w!("{ty_vid:?}")
             }
         }
     }

--- a/crates/flux-refineck/src/type_env.rs
+++ b/crates/flux-refineck/src/type_env.rs
@@ -435,6 +435,9 @@ impl BasicBlockEnvShape {
                 debug_assert!(!scope.has_free_vars(bty));
                 bty.clone()
             }
+            BaseTy::Infer(..) => {
+                tracked_span_bug!("unexpected infer type")
+            }
         }
     }
 

--- a/tests/tests/pos/surface/issue-837.rs
+++ b/tests/tests/pos/surface/issue-837.rs
@@ -1,0 +1,6 @@
+use flux_rs::*;
+
+#[generics(Self as base)]
+trait Trait {}
+
+fn test00(x: &dyn Trait) {}

